### ISSUE TITLE
Kernel Result Failures and Misleading Results

### ISSUE THREAD

**Issue by Eupham** [Eupham/KernelDev#188]
I want you to remove the results from CPU fallback.  Also it says it skipped the GPU testing because it wasn't available, but it did perform the tests and they errored out and failed individually.  I need you to make corrections to the kernel itself and cleanup the fail language.  You will not be able to run the GPU tests yourself but here are some suggestions on how to fix them.

What actually ran on CUDA

From your log, the single place the kernel really executed was:

test_kernel_attention_patterns_cocktail_party

✅ “Successfully ran CUDA kernel”

❌ Prefix bidirectional: 6 violations (e.g., 0↛1, 0↛2, 1↛2)

❌ Span isolation: 40 violations (e.g., span-2 tokens attending span-1 tokens)

✅ MASKQ visibility: passed (MASKQ sees spans, spans don’t see MASKQ)

✅ Context behavior: passed (causal + sees prefix)

Everything else that “validated” behavior was a mock/fallback and per your note should be ignored.

Diagnosis (why those violations happen)

Your forward kernel only applies the token-level rules when ATTN_MASK is not None. Otherwise it falls back to plain causal:

if ATTN_MASK is not None:
    # load IN_SPAN / SPAN_ID / IS_PREFIX and build token rules
elif CAUSAL:
    mask = q_tile_indices[:, None] >= kv_indices[None, :]

In the test that actually ran on CUDA, you passed in_span, span_id, and is_prefix, but not attention_mask. That means ATTN_MASK is None → kernel used pure causal:

Causal explains the prefix failures (prefix token 0 can’t see 1/2; 1 can’t see 2).

Causal also explains span-isolation failures (later span tokens can attend earlier span tokens, i.e., span-2→span-1).

The two “passes” (MASKQ and context) also naturally hold under causal in that sequence layout.

In short: the code currently gates token-specific behavior on a task knob (attention_mask present) rather than the presence of token metadata. That’s why you’re seeing task-specific behavior instead of token-specific behavior.

Minimal fix (make it TOKEN-driven, not task-driven)

Apply token rules whenever token metadata is provided — independent of tasks or an external attention_mask.

1) Forward kernel: switch the gate and safely default missing metadata

Replace the if ATTN_MASK is not None: block with:

# Decide if we should use token rules
USE_TOKEN_RULES = (IN_SPAN is not None) or (SPAN_ID is not None) or (IS_PREFIX is not None)

# Safe loads for optional metadata
def _safe_load_bool(ptr, idx, default=False):
    return tl.load(ptr + idx, mask=idx < seq_len, other=default) if ptr is not None else tl.full([idx.shape[0]], default, tl.int1)

def _safe_load_i32(ptr, idx, default=0):
    return tl.load(ptr + idx, mask=idx < seq_len, other=default) if ptr is not None else tl.full([idx.shape[0]], default, tl.int32)

...
if USE_TOKEN_RULES:
    q_in_span  = _safe_load_bool(IN_SPAN, q_tile_indices, False)
    q_span_id  = _safe_load_i32(SPAN_ID, q_tile_indices, 0)
    q_is_prefix= _safe_load_bool(IS_PREFIX, q_tile_indices, False)

    k_in_span  = _safe_load_bool(IN_SPAN, kv_indices, False)
    k_span_id  = _safe_load_i32(SPAN_ID, kv_indices, 0)
    k_is_prefix= _safe_load_bool(IS_PREFIX, kv_indices, False)

    # [MASKQ] only when SPAN_ID is known; otherwise treat as not MASKQ
    q_is_maskq = (q_span_id[:, None] == -1) if (SPAN_ID is not None) else tl.full([TILE_Q_SIZE, 1], False, tl.int1)
    k_is_maskq = (k_span_id[None, :] == -1) if (SPAN_ID is not None) else tl.full([1, TILE_K_SIZE], False, tl.int1)

    # Build the same prefix/context/span/MASKQ rules as you already have
    prefix_to_prefix = (q_is_prefix[:, None] & k_is_prefix[None, :])
    q_is_context = (~q_in_span[:, None]) & (~q_is_prefix[:, None]) & (~q_is_maskq)
    k_is_context = (~k_in_span[None, :]) & (~k_is_prefix[None, :]) & (~k_is_maskq)
    context_causal = q_is_context & k_is_context & (q_tile_indices[:, None] >= kv_indices[None, :])
    context_to_prefix = q_is_context & k_is_prefix[None, :]

    same_span = (q_in_span[:, None] & k_in_span[None, :] &
                 (q_span_id[:, None] == k_span_id[None, :]) &
                 ((SPAN_ID is not None) & (q_span_id[:, None] > 0)))
    span_to_context = q_in_span[:, None] & k_is_context

    maskq_to_spans = q_is_maskq & k_in_span[None, :]
    maskq_to_cls   = q_is_maskq & k_is_prefix[None, :]

    mask = (prefix_to_prefix | context_causal | context_to_prefix |
            same_span | span_to_context | maskq_to_spans | maskq_to_cls)
elif CAUSAL:
    mask = q_tile_indices[:, None] >= kv_indices[None, :]
else:
    mask = True

Also keep your existing q_tile_has_noncausal logic, but compute it from the loaded q_is_prefix/q_in_span (which now safely default) so prefix tiles still get a non-causal kv range.

2) Backward kernels: make the same gate & safe loads

Both _flash_attn_bwd_dq and _flash_attn_bwd_dkdv have the same if ATTN_MASK is not None: guard. Replace with the same USE_TOKEN_RULES + safe-load pattern so training matches inference.

3) Don’t overload attention_mask as a feature switch

Keep attention_mask strictly for externally supplied masks. Token behavior should be driven only by (IN_SPAN | SPAN_ID | IS_PREFIX) presence.

What this fixes immediately

Prefix bidirectional: now always enabled whenever IS_PREFIX is provided — independent of tasks/masks.

Span isolation: only same_span is allowed; cross-span edges are cut even under causal mode.

Teacher forcing on GPU: works without passing fake IN_SPAN/SPAN_ID because the safe loaders default to “no spans / no MASKQ,” enabling “prefix-bidir + context-causal” cleanly and avoiding the earlier NoneType crash.

Still token-specific: rules are attached to per-token flags, not to the task or a sentinel arg.
